cpp/include/cudf/io/detail/data_casting.cuh

```diff
@@ -128,20 +127,24 @@ __device__ __forceinline__ int32_t parse_unicode_hex(char const* str)
 }
 
 /**
- * @brief Writes the UTF-8 byte sequence to \p out_it and returns the iterator to one past the
- * last item that was written to \p out_it
+ * @brief Writes the UTF-8 byte sequence to \p out_it and returns the number of bytes written to
+ * \p out_it
  */
-template <typename utf8_char_t, typename out_it_t>
-__device__ __forceinline__ out_it_t write_utf8_char(utf8_char_t utf8_chars, out_it_t out_it)
+template <typename utf8_char_t>
+__device__ __forceinline__ size_type write_utf8_char(utf8_char_t utf8_chars, char*& out_it)
 {
   constexpr size_type MAX_UTF8_BYTES_PER_CODE_POINT = 4;
   char char_bytes[MAX_UTF8_BYTES_PER_CODE_POINT];
   auto const num_chars_written = strings::detail::from_char_utf8(utf8_chars, char_bytes);
+  size_type bytes              = 0;
 
   for (size_type i = 0; i < MAX_UTF8_BYTES_PER_CODE_POINT; i++) {
-    if (i < num_chars_written) { *out_it++ = char_bytes[i]; }
+    if (i < num_chars_written) {
+      if (out_it) *out_it++ = char_bytes[i];
+      ++bytes;
+    }
   }
-  return out_it;
+  return bytes;
```

Comment:
When is `bytes != num_chars_written`?
This function seems overly complex. Maybe I'm missing something?
```
constexpr size_type write_utf8_char(char_utf8 character, char*& out_it) {
   auto const bytes = (out_it == nullptr) ? strings::detail::bytes_in_char_utf8(character) : 
                                            strings::detail::from_char_utf8(character, out_it);
   out_it += bytes;
   return bytes;
}
```
There does not seem to be a need for the template as well.

Reply:
I think, this is written to allow the compiler to unroll the loop.
I will change this to above suggestion.